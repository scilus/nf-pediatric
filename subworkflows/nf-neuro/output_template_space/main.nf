include { UTILS_TEMPLATEFLOW                    } from '../../../modules/nf-neuro/utils/templateflow/main.nf'
include { BETCROP_FSLBETCROP as BET_T1W         } from '../../../modules/nf-neuro/betcrop/fslbetcrop/main.nf'
include { BETCROP_FSLBETCROP as BET_T2W         } from '../../../modules/nf-neuro/betcrop/fslbetcrop/main.nf'
include { REGISTRATION_ANTS                     } from '../../../modules/nf-neuro/registration/ants/main.nf'
include { REGISTRATION_ANTSAPPLYTRANSFORMS      } from '../../../modules/nf-neuro/registration/antsapplytransforms/main.nf'
include { REGISTRATION_TRACTOGRAM               } from '../../../modules/nf-neuro/registration/tractogram/main.nf'

workflow OUTPUT_TEMPLATE_SPACE {

    take:
        ch_anat // channel: [ val(meta), [ anat ] ]
        ch_nifti_files // channel: [ val(meta), [ nifti_files ] ]
        ch_trk_files // channel: [ val(meta), [ trk_files ] ]

    main:

    ch_versions = Channel.empty()

    // ** First, let's assess if the desired template exists in      ** //
    // ** the templateflow home directory (user-specified as params. ** //
    // ** or default to $outdir/../templateflow)                     ** //
    if ( !file("${params.templateflow_home}/tpl-${params.template}").exists() ) {
        log.info("Template ${params.template} not found in " +
                "${params.templateflow_home}. Will be downloaded." +
                "If you do not have access to the internet while running" +
                "this pipeline, please download the template manually" +
                "and provide the location using --templateflow_home.")
        log.info("${params.template} will be downloaded at resolution " +
                "${params.templateflow_res} from cohort ${params.templateflow_cohort}.")

        UTILS_TEMPLATEFLOW (
            [
                params.template,
                params.templateflow_res != null ? params.templateflow_res : [],
                params.templateflow_cohort != null ? params.templateflow_cohort : []
            ]
        )
        ch_versions = ch_versions.mix(UTILS_TEMPLATEFLOW.out.versions)

        // ** Setting outputs ** //
        ch_t1w_tpl = UTILS_TEMPLATEFLOW.out.T1w
        ch_t2w_tpl = UTILS_TEMPLATEFLOW.out.T2w

    } else {
        // ** If the template exists, we will not download it again. ** //
        log.info("Template ${params.template} found in " +
                "${params.templateflow_home}. Will be used.")

        // ** Load the files from the templateflow directory ** //
        def path = "${params.templateflow_home}/tpl-${params.template}/"
        if ( params.templateflow_cohort ) {
            ch_t1w_tpl = Channel.fromPath(
                "${path}/${params.templateflow_cohort}/*res-*${params.templateflow_res}_T1w.nii.gz",
                checkIfExists: false
            )
            ch_t2w_tpl = Channel.fromPath(
                "${path}/${params.templateflow_cohort}/*res-*${params.templateflow_res}_T2w.nii.gz",
                checkIfExists: false
            )
        } else {
            ch_t1w_tpl = Channel.fromPath(
                "${path}/*res-*${params.templateflow_res}_T1w.nii.gz",
                checkIfExists: false
            )
            ch_t2w_tpl = Channel.fromPath(
                "${path}/*res-*${params.templateflow_res}_T2w.nii.gz",
                checkIfExists: false
            )
        }
    }

    // ** The template may not have a brain mask, so we will ** //
    // ** run BET by default (bit painful, but necessary)    ** //
    ch_bet_tpl_t1w = ch_t1w_tpl
        | map{ t1w -> [ [id: "template"], t1w, [], [] ] }

    BET_T1W ( ch_bet_tpl_t1w )
    ch_versions = ch_versions.mix(BET_T1W.out.versions)
    // ** Strip the template from the meta field so we can combine it ** //
    ch_t1w_tpl = BET_T1W.out.image
        | map{ meta, image -> image }

    ch_bet_tpl_t2w = ch_t2w_tpl
        | map{ t2w -> [ [id: "template"], t2w, [], [] ] }

    BET_T2W ( ch_bet_tpl_t2w )
    ch_versions = ch_versions.mix(BET_T2W.out.versions)
    // ** Strip the template from the meta field so we can combine it ** //
    ch_t2w_tpl = BET_T2W.out.image
        | map{ meta, image -> image }

    // ** Register the subject to the template space ** //
    ch_registration = ch_anat
        | combine(params.use_template_t2w ? ch_t2w_tpl : ch_t1w_tpl)
        | map{ meta, anat, tpl -> tuple(meta, anat, tpl, []) }

    REGISTRATION_ANTS ( ch_registration )
    ch_versions = ch_versions.mix(REGISTRATION_ANTS.out.versions)

    // ** Apply the transformation to all files ** //
    // ** The channel ch_nifti_files contains all the files that ** //
    // ** need to be transformed to the template space in that structure: ** //
    // ** [ tuple(meta, [ file1, file2, ... ]) ] ** //
    // ** Need to unpack the files and apply the transformation to each one ** //
    ch_files_to_transform = ch_nifti_files
        | join(REGISTRATION_ANTS.out.image)
        | join(REGISTRATION_ANTS.out.warp)
        | join(REGISTRATION_ANTS.out.affine)

    REGISTRATION_ANTSAPPLYTRANSFORMS ( ch_files_to_transform )
    ch_versions = ch_versions.mix(REGISTRATION_ANTSAPPLYTRANSFORMS.out.versions)

    // ** Apply the transformation to the tractograms ** //
    ch_tractograms_to_transform = ch_trk_files
        | join(REGISTRATION_ANTS.out.image)
        | join(REGISTRATION_ANTS.out.warp)
        | join(REGISTRATION_ANTS.out.affine)
        | map{ meta, trk, image, warp, affine ->
            tuple(meta, image, affine, trk, [], warp)
        }

    REGISTRATION_TRACTOGRAM ( ch_tractograms_to_transform )
    ch_versions = ch_versions.mix(REGISTRATION_TRACTOGRAM.out.versions)

    emit:
        versions = ch_versions                     // channel: [ versions.yml ]
}

