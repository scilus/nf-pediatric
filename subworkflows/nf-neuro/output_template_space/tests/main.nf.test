nextflow_workflow {

    name "Test Subworkflow OUTPUT_TEMPLATE_SPACE"
    script "../main.nf"
    workflow "OUTPUT_TEMPLATE_SPACE"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/output_template_space"

    tag "registration"
    tag "registration/ants"
    tag "registration/antsapplytransforms"
    tag "registration/tractogram"
    tag "image"
    tag "image/applymask"
    tag "betcrop"
    tag "betcrop/fslbetcrop"
    tag "utils"
    tag "utils/templateflow"
    tag "load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "tractometry.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    } /* Commented since we do not have local data on runners
    test("output to template MNI152NLin2009cAsym with local folder") {
        config "./nextflow_local.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        tractometry: it.simpleName == "tractometry"
                    }
                input[0] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        file("\${test_data_directory}/mni_masked.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.nii.gz")
                        ]
                    ]
                }
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM_labels_map.nii.gz")
                        ]
                    ]
                }
                input[4] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.trk"),
                            file("\${test_data_directory}/IFGWM_color.trk"),
                            file("\${test_data_directory}/IFGWM_uni.trk")
                        ]
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    } */
    test("output to template MNI152NLin2009aAsym - with brain mask") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        tractometry: it.simpleName == "tractometry"
                    }
                input[0] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        file("\${test_data_directory}/mni_masked.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.nii.gz")
                        ]
                    ]
                }
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM_labels_map.nii.gz")
                        ]
                    ]
                }
                input[4] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.trk"),
                            file("\${test_data_directory}/IFGWM_color.trk"),
                            file("\${test_data_directory}/IFGWM_uni.trk")
                        ]
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
    test("output to template MNI152NLin2009aAsym - stub") {
        config "./nextflow.config"
        options "-stub-run"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        tractometry: it.simpleName == "tractometry"
                    }
                input[0] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        file("\${test_data_directory}/mni_masked.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.nii.gz")
                        ]
                    ]
                }
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM_labels_map.nii.gz")
                        ]
                    ]
                }
                input[4] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.trk"),
                            file("\${test_data_directory}/IFGWM_color.trk"),
                            file("\${test_data_directory}/IFGWM_uni.trk")
                        ]
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out.versions).match()}
            )
        }
    }
}
