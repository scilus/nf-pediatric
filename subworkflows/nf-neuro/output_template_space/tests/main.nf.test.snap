{
    "output to template MNI152NLin2009aAsym - with brain mask": {
        "content": [
            {
                "0": [
                    "template__image_bet.nii.gz:md5,01c9075c8bf0c0f177660e0235dd6dec"
                ],
                "1": [
                    "template__image_bet.nii.gz:md5,a86fa91b93732a6a8bb0ef7ee8c6f758"
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,b99092deb15615bc97ea03eacccff930"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test__IFGWM__warped.nii.gz:md5,fb770a7e718011de4da6b23a712fcb03"
                    ]
                ],
                "4": [
                    
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test__IFGWM_labels_map__warped.nii.gz:md5,d31f6ce5792d9ceeae93d987f6fa82eb"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test__IFGWM.trk:md5,33348cef01be731c98628826e0a57649",
                            "test__IFGWM_color.trk:md5,33348cef01be731c98628826e0a57649",
                            "test__IFGWM_uni.trk:md5,a0af47ad2847103505ebfb13c14b1d7d"
                        ]
                    ]
                ],
                "7": [
                    "versions.yml:md5,04f69537f2d6cdc20e0a8c110ae9b7e7",
                    "versions.yml:md5,0d37dad46b915d0f062b1e1c791f47a7",
                    "versions.yml:md5,29c5499c53451b27652c9eb978a79e76",
                    "versions.yml:md5,75c8524f95824b358ede7fe52dfa35e6",
                    "versions.yml:md5,a6b014172fd4dfe073bb793b58943c46",
                    "versions.yml:md5,ad180f312de6e9124ce2789c69d3afdf",
                    "versions.yml:md5,b180b92a88347fe68d25e2c8066322f1"
                ],
                "ch_registered_anat": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,b99092deb15615bc97ea03eacccff930"
                    ]
                ],
                "ch_t1w_tpl": [
                    "template__image_bet.nii.gz:md5,01c9075c8bf0c0f177660e0235dd6dec"
                ],
                "ch_t2w_tpl": [
                    "template__image_bet.nii.gz:md5,a86fa91b93732a6a8bb0ef7ee8c6f758"
                ],
                "ch_warped_labels_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test__IFGWM_labels_map__warped.nii.gz:md5,d31f6ce5792d9ceeae93d987f6fa82eb"
                    ]
                ],
                "ch_warped_mask_files": [
                    
                ],
                "ch_warped_nifti_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test__IFGWM__warped.nii.gz:md5,fb770a7e718011de4da6b23a712fcb03"
                    ]
                ],
                "ch_warped_trk_files": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test__IFGWM.trk:md5,33348cef01be731c98628826e0a57649",
                            "test__IFGWM_color.trk:md5,33348cef01be731c98628826e0a57649",
                            "test__IFGWM_uni.trk:md5,a0af47ad2847103505ebfb13c14b1d7d"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,04f69537f2d6cdc20e0a8c110ae9b7e7",
                    "versions.yml:md5,0d37dad46b915d0f062b1e1c791f47a7",
                    "versions.yml:md5,29c5499c53451b27652c9eb978a79e76",
                    "versions.yml:md5,75c8524f95824b358ede7fe52dfa35e6",
                    "versions.yml:md5,a6b014172fd4dfe073bb793b58943c46",
                    "versions.yml:md5,ad180f312de6e9124ce2789c69d3afdf",
                    "versions.yml:md5,b180b92a88347fe68d25e2c8066322f1"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-29T13:56:29.991808"
    },
    "output to template MNI152NLin2009aAsym - stub": {
        "content": [
            [
                "versions.yml:md5,04f69537f2d6cdc20e0a8c110ae9b7e7",
                "versions.yml:md5,08c023826481c9130c3914096cda7f63",
                "versions.yml:md5,0d37dad46b915d0f062b1e1c791f47a7",
                "versions.yml:md5,a166b69e9dc7963b29c7a5944fbbf72a",
                "versions.yml:md5,a6b014172fd4dfe073bb793b58943c46",
                "versions.yml:md5,ad180f312de6e9124ce2789c69d3afdf",
                "versions.yml:md5,d4da7406446940901e5757ec63bf3d8f"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-29T13:57:01.689815"
    }
}